[Flang] Move builtin .mod generation into runtimes (Reapply #137828)

Meinersbur opened the pull request:
Reapplication of #137828, changes:
 * Workaround CMAKE_Fortran_PREPROCESS_SOURCE issue for CMake < 2.24: The issue is that `try_compile` does not forward manually-defined compiler flang variables to the test build environment; instead of just a negative test result, it aborts the configuration step itself. To be fair, manually defining these variables is deprecated since at least CMake 3.6.
 * Missing flang cmd line flags for CMake < 3.28 `-target=`, `-O2`, `-O3`
 * It is now possible to set FLANG_RT_ENABLED_STATIC=OFF and FLANG_RT_ENABLE_SHARED=OFF at the same and is the default for amdgpu and nvptx targets. In this mode, only the .mod files are compiled -- necessary for module files in lib/clang/22/finclude/flang/(nvptx64-nvidia-cuda|amdgpu-amd-amdhsa)/*.mod to be available.
 * For compiling omp_lib.mod for nvptx and amdgpu, the module build functionality must be hoisted out if openmp's runtime/ directory which is only included for host targets. This PR now requires #169909.
 

Move building the .mod files from openmp/flang to openmp/flang-rt using a shared mechanism. Motivations to do so are:

1. Most modules are target-dependent and need to be re-compiled for each target separately, which is something the LLVM_ENABLE_RUNTIMES system already does. Prime example is `iso_c_binding.mod` which encodes the target's ABI. Constants such as [`c_long_double` also have different values](https://github.com/llvm/llvm-project/blob/d748c81218bee39dafb9cc0c00ed7831a3ed44c3/flang-rt/lib/runtime/iso_c_binding.f90#L77-L81). Most other modules have `#ifdef`-enclosed code as well. For instance this caused offload targets nvptx64-nvidia-cuda/amdgpu-amd-amdhsa to use the modules files compiled for the host which may contrain uses of the types REAL(10) or REAL(16) not available for nvptx/amdgpu.

#146876
#128015
#129742
#158790

3. CMake has support for Fortran that we should use. Among other things, it automatically determines module dependencies so there is no need to hardcode them in the CMakeLists.txt.

4. It allows using Fortran itself to implement Flang-RT. Currently, only `iso_fortran_env_impl.f90` emits object files that are needed by Fortran applications (#89403). The workaround of #95388 could be reverted (PR #169525).


If using Flang for cross-compilation or target-offloading, flang-rt must now be compiled for each target not only for the library, but also to get the target-specific module files. For instance in a bootstrapping runtime build, this can be done by adding: `-DLLVM_RUNTIME_TARGETS=default;nvptx64-nvidia-cuda;amdgpu-amd-amdhsa`.


Some new dependencies come into play:
 * openmp depends on flang-rt for building `lib_omp.mod` and `lib_omp_kinds.mod`. Currently, if flang-rt is not found then the modules are not built.
 * check-flang depends on flang-rt: If not found, the majority of tests are disabled. If not building in a bootstrpping build, the location of the module files can be pointed to using `-DFLANG_INTRINSIC_MODULES_DIR=<path>`, e.g. in a flang-standalone build. Alternatively, the test needing any of the intrinsic modules could be marked with `REQUIRES: flangrt-modules`.
 * check-flang depends on openmp: Not a change; tests requiring `lib_omp.mod` and `lib_omp_kinds.mod` those are already marked with `openmp_runtime`.

As intrinsic are now specific to the target, their location is moved from `include/flang` to `<resource-dir>/finclude/flang/<triple>`. The mechnism to compute the location have been moved from flang-rt (previously used to compute the location of `libflang_rt.*.a`) to common locations in `cmake/GetToolchainDirs.cmake` and `runtimes/CMakeLists.txt` so they can be used by both, openmp and flang-rt. Potentially the mechnism could also be shared by other libraries such as compiler-rt. 

`finclude` was chosen because `gfortran` uses it as well and avoids misuse such as `#include <flang/iso_c_binding.mod>`. The search location is now determined by `ToolChain` in the driver, instead of by the frontend. Another subdirectory `flang` avoid accidental inclusion of gfortran-modules which due to compression would result in user-unfriendly errors. Now the driver adds `-fintrinsic-module-path` for that location to the frontend call (Just like gfortran does). `-fintrinsic-module-path` had to be fixed for this because ironically it was only added to `searchDirectories`, but not `intrinsicModuleDirectories_`. Since the driver determines the location, tests invoking `flang -fc1` and `bbc` must also be passed the location by llvm-lit. This works like llvm-lit does for finding the include dirs for Clang using `-print-file-name=...`.

Meinersbur committed:
Reapply "[Flang] Move builtin .mod generation into runtimes" (#169489)

This reverts commit 4bc654d6497430c1dd5e4e25aaa84b3dec3e1113.

Meinersbur committed:
Clarify try_compile isolation

Meinersbur committed:
Prefer -J over -module-dir

Meinersbur committed:
Add modules-only mode

Meinersbur committed:
Fix remaining include paths

Meinersbur commented:
@jhuber6 Based on the feedback in #137828, I made it easier to build the Fortran modules for nvptx and amdgpu targets. It is now possible to set both FLANG_RT_ENABLE_STATIC and FLANG_RT_ENABLE_SHARED to off, meaning no library is build, only the modules files. This is the default if the target is nvptx or amdgpu. However, this overrides the (experimental) default behavior of #131826. I could not get the gpu_sources to compile since it cannot find include files such as `#include <cassert>`. I could trick it to point to the libc++ versions, but I don't want other users have to do many steps. Please have a look specifically to [`9041b78` (#169638)](https://github.com/llvm/llvm-project/pull/169638/commits/9041b784e3c054266a9109c2698030ea4f2aad44)

In the spirit off adding `LLVM_ENABLE_RUNTIMES=flang-rt` automatically, we could also automatically add `LLVM_RUNTIME_TARGETS=default;amdgcn-amd-amdhsa` when the AMDGPU backend is enabled; respectively `LLVM_RUNTIME_TARGETS=default;nvptx64-nvidia-cuda` if `LLVM_TARGETS_TO_BUILD=NVPTX` is present. What do you think?

Meinersbur committed:
Fix LIBOMP_COPY_EXPORTS

Meinersbur committed:
Simplify by not using macro

vzakhari commented on flang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1568,6 +1558,14 @@ bool CompilerInvocation::createFromArgs(
+          args.getLastArg(clang::options::OPT_resource_dir))
+    invoc.resourceDir = a->getValue();
+  else
+    invoc.resourceDir = clang::driver::Driver::GetResourcesPath(
```
I've just started trying this patch, and I see a build failure suggesting that this should be `clang::GetResourcesPath`.

Meinersbur replied:
There have been several reverts and re-applies of this patch: https://github.com/llvm/llvm-project/pull/169397. The pre-merge CI says it is fine: https://github.com/llvm/llvm-project/actions/runs/19701948097/job/56440119613?pr=169638 on the base patch as-is.

Meinersbur replied:
Merged-in the latest main which includes the re-land (https://github.com/llvm/llvm-project/commit/d090311aa7df7cbd93412c31685893aefc4ad341). Should resole the issue unless reverted again.

Meinersbur committed:
Solve CMAKE_Fortran_PREPROCESS_SOURCE issue

Meinersbur committed:
Add TODO comment

Meinersbur committed:
Merge remote-tracking branch 'official/main' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
Integrate HandleFortran again

vzakhari commented:
Thank you, Michael! The patchs seems to be working on my side (after some adjustments downstream). So it is good to go for the next week merge (once you merge it, I will resolve conflicts on my side).

I do not know about the experimental direct GPU build. @jhuber6 can you please comment on Michael's question above?

Meinersbur committed:
Build omp_lib.mod on device

Meinersbur committed:
Fix modules-only condition

Meinersbur committed:
Backport openmp

Meinersbur committed:
Use current .mod building mechanism

Meinersbur committed:
Allow non-LIBOMP_FORTRAN_MODULES mode

Meinersbur committed:
Also move omp_lib.h

Meinersbur committed:
CMake-ify

Meinersbur committed:
Diagnostics

Meinersbur committed:
Reuse original file location where check-flang expects them

Meinersbur committed:
Fix DEPENDS

Meinersbur committed:
Merge branch 'users/meinersbur/openmp_hoist-module' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
LIBOMP_COPY_EXPORT from original location

Meinersbur committed:
Merge branch 'users/meinersbur/openmp_hoist-module' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Re-add LIBOMP_FORTRAN_MODULES visible option

Meinersbur committed:
Post-merge fixes

Meinersbur committed:
Exact CMake version

Meinersbur committed:
Merge branch 'users/meinersbur/openmp_hoist-module' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Skip libraries other than runtime on module-only mode

Meinersbur committed:
Always create libomp-mod target for robustness

Meinersbur committed:
Replace LIBOMP_FORTRAN_MODULES with RUNTIMES_FLANG_MODULES_ENABLED

For some reason there were to different flags to build the Fortran modules:

1. LIBOMP_FORTRAN_MODULES=ON and CMAKE_Fortran_COMPILER

2. LIBOMP_FORTRAN_MODULES_COMPILER

When using the second, LIBOMP_FORTRAN_MODULES would be off. Trying to
keep the option available causes build failures by misinterpretation.

Meinersbur committed:
Merge branch 'main' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Fix disabled flang_rt.quad.math

Meinersbur committed:
Merge remote-tracking branch 'official/main' into users/meinersbur/openmp_hoist-module

Meinersbur committed:
Merge branch 'users/meinersbur/openmp_hoist-module' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Fix oversights

Meinersbur committed:
Fix build with f128_sources

kkwli commented:
On AIX, there is only `iso_fortran_env_impl.mod`.
```
$ find . -name '*.mod'
./flang-rt/lib/runtime/iso_fortran_env_impl.mod
```
On Linux, I can see all the modules generated.
```
$ find . -name '*.mod'
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/__ppc_types.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/__fortran_builtins.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/__cuda_builtins.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/__fortran_type_info.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/iso_c_binding.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/iso_fortran_env_impl.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/mma.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/ieee_features.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/__cuda_device.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/__fortran_ieee_exceptions.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/__ppc_intrinsics.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/iso_fortran_env.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/cudadevice.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/ieee_exceptions.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/ieee_arithmetic.mod
./lib/clang/22/finclude/flang/powerpc64le-unknown-linux-gnu/cooperative_groups.mod
```

DanielCChen reviewed (approved):
LGTM.
Thanks!

kkwli reviewed (approved):
LG. Thanks.

Meinersbur committed:
Merge commit '4b2714f12f592ef3237056759dbe735d43fbeec1^' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Merge commit '4b2714f12f592ef3237056759dbe735d43fbeec1' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Merge commit '830ef4e4e2d0f181ed1cfacbc3daacb2555ff9e4^' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Merge commit '830ef4e4e2d0f181ed1cfacbc3daacb2555ff9e4' into users/meinersbur/flang_builtin-mods_2

Meinersbur committed:
Post-merge fixes

petrhosek commented:
This broke the runtimes build on Windows when targeting a non-Windows target with the following error:
```
CMake Error at C:/b/s/w/ir/x/w/llvm-llvm-project/cmake/Modules/GetToolchainDirs.cmake:76 (string):
  string sub-command REPLACE requires at least four arguments.
Call Stack (most recent call first):
  C:/b/s/w/ir/x/w/llvm-llvm-project/cmake/Modules/GetToolchainDirs.cmake:42 (get_toolchain_arch_dirname)
  CMakeLists.txt:459 (get_toolchain_library_subdir)
```
Would it be possible to revert the change while this is being investigated?

Since this makes significant changes to the runtimes build that affects non-Flang users, I'd also appreciate to be included as a reviewer on the reland.
